File: python/src/partition_tree.py
#!/usr/bin/env python3


def get_size_tree(partition_class_tree, partition_exists_pred):
    '''
    Returns a tree of concrete partitions,
    NOT partition classes.
    '''
    leaf = (None, ())

    def size(n):
        print(n)
        (s, idx), cs = n
        return s

    def _get_size_tree(node, top_idx):
        partition_class, children = node
        children_results = ()
        if children != (leaf, ):
            for i in range(partition_class.max_idx_value()):
                new_pc = children[partition_class.idx_to_child_kind(i)]
                new_idx = top_idx + (i, )
                # The predicate needs not the first index
                if partition_exists_pred(new_idx[1:]):
                    r = _get_size_tree(new_pc, new_idx)
                    if size(r) > 0:
                        children_results += (r, )
        else:
            return (partition_class.max_idx_value(), top_idx[-1]), ()

        total_size = sum(size(cr) for cr in children_results)

        return (total_size, top_idx[-1]), children_results

    return _get_size_tree(partition_class_tree, (None, ))


def get_start_tree(size_tree):
    '''
    Depth-first allocation order.
    For other allocation orders,
    swap levels in the tree.
    '''
    def _start_tree(size_tree, start):
        (size, idx), children = size_tree
        starts = tuple(start + sum(s for (s, _1), _2 in children[:i])
                       for i in range(len(children)))
        print("STARTS", starts, "START", start)
        new_children = tuple(
            _start_tree(c, s) for c, s in zip(children, starts))
        return (start, idx), new_children

    return _start_tree(size_tree, 0)


def get_start_idx(size_tree):
    '''
    Return first (depth-first-wise) indices list
    in a tree.
    '''
    (s, idx), children = size_tree

    def get_idx(subtree):
        (_1, idx), _2 = subtree
        return idx

    if children == ():
        return (0, )
    else:
        return (get_idx(children[0]), ) + get_start_idx(children[0])


def iterate(size_tree, idxs):
    '''
    `ptree` is a tree of concrete partitions,
    where each node is in the form
    '''
    def get_idx(subtree):
        (_1, idx), _2 = subtree
        return idx

    def _iterate(size_tree, idxs):
        (s, idx), children = size_tree
        nidx = idxs[0]

        if children == ():  # Leaf case
            new_idx = nidx + 1
            if new_idx == s:
                return None
            else:
                return (nidx + 1, )
        else:  # Node case
            # looking for child with the right idx
            i, c = next(
                (i, c) for i, c in enumerate(children) if get_idx(c) == nidx)

            sub_idxs = _iterate(c, idxs[1:])
            if sub_idxs is None:  # We need to move to the next child
                if i == len(children) - 1:  # We give up
                    return None
                else:
                    sub_idxs = get_start_idx(children[i + 1])
                    return (i + 1, ) + sub_idxs
            else:
                return (i, ) + sub_idxs

    return _iterate(size_tree, idxs)

File: cpp/src/partition_tree.cpp
#include "partition_tree.hpp"
#include <memory>
namespace hypercubes {
namespace slow {

TreeP<std::pair<int, int>> get_size_tree(const PartitionTree &t,
                                         PartitionPredicate predicate) {

  auto binarified_predicate = [&predicate](Indices idx) {
    return to_bool(predicate(idx));
  };
  return get_size_tree(t, binarified_predicate);
}

TreeP<std::pair<int, int>> get_size_tree(const PartitionTree &t,
                                         std::function<bool(Indices)> predicate)

{

  using ResType = TreeP<std::pair<int, int>>;

  using F = std::function<ResType(const PartitionTree &, //
                                  const Indices &)>;

  auto getsize = [](ResType t) { return t->n.first; };

  F _get_size_tree = [&predicate,      //
                      &_get_size_tree, //
                      &getsize](const PartitionTree &pct,
                                const Indices &top_idxs) -> ResType {
    auto partition_class = pct->n;
    auto children = pct->children;
    vector<ResType> children_results;
    int nodesize = 0;
    if (children.size() != 0) {
      for (int idx = 0; idx < partition_class->max_idx_value(); ++idx) {
        auto new_pc = children[idx];
        Indices new_idxs = top_idxs;
        Indices predicate_idxs;
        new_idxs.push_back(idx);
        std::copy(new_idxs.begin() + 1, new_idxs.end(),
                  std::back_inserter(predicate_idxs));
        if (predicate(predicate_idxs)) {
          ResType r = _get_size_tree(new_pc, new_idxs);
          if (getsize(r) != 0)
            children_results.push_back(r);
        }
      }
      for (const auto &c : children_results)
        nodesize += getsize(c);
    } else
      nodesize = partition_class->max_idx_value();

    auto subn = std::make_pair(nodesize, //
                               *top_idxs.rbegin());
    return mt(subn, children_results);
  };

  return _get_size_tree(t, {0});
}
TreeP<std::pair<int, int>>
get_offset_tree(const TreeP<std::pair<int, int>> &size_tree) {
  using SizeTree = TreeP<std::pair<int, int>>;
  using ResType = TreeP<std::pair<int, int>>;

  using F = std::function<ResType(const SizeTree &, //
                                  int)>;
  F _get_start_tree = [&_get_start_tree](const SizeTree &st, //
                                         int start) -> ResType {
    vector<ResType> children_results;
    int _idx_so_far = start;
    for (const auto &child : st->children) {
      children_results.push_back(_get_start_tree(child, _idx_so_far));
      _idx_so_far += child->n.first;
    }
    return mt(std::make_pair(start, st->n.second), children_results);
  };
  return _get_start_tree(size_tree, 0);
}

} // namespace slow
} // namespace hypercubes
